accepted/2022/simplify-rid-model.md

```diff
@@ -0,0 +1,244 @@
+- Implement a new algorithmic host scheme for RID selection, enabled by default.
+
+The new scheme will be based on the following:
+
+- Each host is built with a hard-coded set of RIDs that it probes for, both triplets and singles.
+- Each host is already built for a specific RID triple, so this change is an evolutionary change.
+- There are processor-agnostic managed-code scenarios where a RID single is relevant, like Windows-only code (for example, accessing the registry) that works the same x64, Arm64 and x86. The same is true for macOS and Linux.
+- The RID that `dotnet --info` returns will always match the first in this hard-coded set of RIDs.
+
+Let's assume that an app is running on Alpine 3.16, the host will probe for the following RIDs:
+
+- `linux-musl-x64`
+- `unix`
+
+Ubuntu would be similar:
+
+- `linux-arm64`
+- `unix`
+
+macOS would be similar:
+
+- `osx-arm64`
+- `unix`
+
+Windows would similar:
+
+- `win-x86`
+- `win`
+
+Note: `win7` and `win10` also exist (as processor-specific and -agnostic). Ideally, we don't have to support those in the host-specific scheme, but we need to do some research on that.
+
+Note: A mix of processor-specific RIDs are used above, purely to demonstrate the range of processor-types supported. Each operating system supports a variety of processor-types, and this is codified in their associated RIDs.
+
+More generally, the host will probe for:
+
+- This environment, RID triplet (OS-CRuntime-arch)
+- This environment, RID single (OS-only).
+
+Note: The abstract `unix` RID is used for macOS and Linux, instead of a concrete RID for those OSes.
+
+The host will implement "first one wins" probing logic for selecting a RID-specific asset.
+
+This behavior only applies for portable apps. RID-specific apps will not probe for RID-specific assets.
+
+There are other RIDs in `runtimes.json` today. This new scheme will not support those. The `runtimes.json` host compat feature will need to be used to support those, which is likely very uncommon. Even if they are quasi-common, we will likely still make this change and require some combination of the ecosystem to adapt and app developers to opt-in to the `runtimes.json` compat mode.
+
+The host and `runtimes.json` must remain compatible. The RIDs known by the host must be a subset of `runtimes.json` RIDs. We may need to update `runtimes.json` to ensure that the RIDs known by the host are present in that file.
+
+
+## Minimum CRT/libc version
+
+This scheme doesn't explicitly define a CRT/libc version. Instead, we will define a minimum CRT/libc version, per major .NET version. That minimum CRT/libc version will form a contract and should be documented, per .NET version. Practically, it defines the oldest OS version that you can run an app on, for a given .NET version.
+
+For .NET 7:
+
+- For Windows, .NET will target the Windows 10 CRT.
+- For Linux with glibc, .NET will target CentOS 7.
```

Comment:
Should this be codified to a concrete glibc version (e.g., 2.17 that is used on CentOS 7)) rather than CentOS version itself to make this clear for other distros?

Reply:
Yes and no. I think that's somewhat confusing the point. My understanding is that we're targeting the glibc that CentOS 7 has, whatever it is, making the specific version irrelevant. Fair?

Reply:
A little more context. We target the glibc in CentOS 7 to build our `linux-x64` product. If there are distros with an older glibc, then we don't support them, by definition. If .NET is run on a distro with a same/newer glibc, then it is all good. If you build .NET yourself on a distro with a different glibc version, that doesn't matter. It has nothing to do with this plan.

Reply:
I agree with @tkapin that this should be speced using the glibc version. A few reasons:
- If I have some random Linux distro with given glibc version, I should be able to tell whether it is compatible without downloading CentOS 7 just to find the glibc version that happens to be in it. We should not take favorites with the distros.
- The reference distro used by Microsoft official build is likely going to change over time.

It is fine for the spec to explain the underlying reason behind the choice (e.g. that it is the version in CentOS that is the oldest among supported mainstream Linux distros currently).

This applies for both glibc and musl.

Reply:
_RHEL/CentOS 7 is so old, there can't be anything older._ :smile: 

It has glibc v2.17.

```
$ podman run centos:7 yum list installed glibc
Loaded plugins: fastestmirror, ovl
Installed Packages
glibc.x86_64                        2.17-317.el7                         @CentOS
```

Reply:
To add one more on top of @jkotas reasons, if we expressed this dependency in terms of distro version, that distro would either have to guarantee that it doesn't offer glibc upgrade to another version within the same version of the distro, or we would have to ensure that we support the earliest version of glibc shipps with that distro version.

We should definitely list the OSes that are supposed to be compatible, but since the primary thing is compatibility with glibc, I think it should be canonically expressed via glibc version with the distro version deduced from it rather than the other way around.

I'm not having an extremely strong opinion about this though so if you feel the former would be more transparent, I'm ok with that.

Reply:
Fixed. You guys are right.